### PR TITLE
ENH: Merge PseudoPositioner children in plans automatically 

### PR DESCRIPTION
**Pull request by tacaswell**
- merge_cycler attempts to find PseudoSingle devices (in a duck-type
   way) that are from the same device and merge them into a single
   cycler entry.
 - after we set up the meta-data in scan_nd, merge the cycler!

### CODE REVIEW

File: bluesky/utils.py [hunk 1]
```diff
@@ -1080,3 +1086,52 @@ def __call__(self, status_objs_or_none):
 def _L2norm(x, y):
     "works on (3, 5) and ((0, 3), (4, 0))"
     return np.sqrt(np.sum((np.asarray(x) - np.asarray(y))**2))
+
+
+def merge_cycler(cyc):
```
Comment (danielballan) [inline bluesky/utils.py]: Suggested docstring:

```
Specify movements of sets of interdependent axes atomically.

Inspect the keys of ``cyc`` (which are Devices) to indentify those which are
interdependent. Merge those speciously independent entries into one entry.
```

and then show a simple example with two interdependent keys and one independent key.

File: bluesky/utils.py [hunk 2]
```diff
@@ -1080,3 +1086,52 @@ def __call__(self, status_objs_or_none):
...
+    def munge_list_of_parts(objs):
+        def get_parent(o):
+            return getattr(o, 'parent')
+
+        independent_objs = set()
+        maybe_coupleed = set()
```
Comment (danielballan) [inline bluesky/utils.py]: maybe_coupleed -> maybe_coupled

File: bluesky/utils.py [hunk 3]
```diff
@@ -1080,3 +1086,52 @@ def __call__(self, status_objs_or_none):
...
+                complex_objs.add(o)
+            elif (parent is not None and hasattr(parent, 'RealPosition')):
+                maybe_coupleed.add(o)
+            else:
+                independent_objs.add(o)
+
+        return (independent_objs, complex_objs,
+                groupby(get_parent, maybe_coupleed))
+
+    def my_name(obj):
```
Comment (danielballan) [inline bluesky/utils.py]: Add comment:

```
# This gets the name of the attribute on `obj.parent` which refer to `obj`.
# This is obj's attribute name, distinct from its `obj.name`.
```

File: bluesky/utils.py [hunk 4]
```diff
@@ -1080,3 +1086,52 @@ def __call__(self, status_objs_or_none):
...
+    # only simple non-coupled objects, declare victory and bail!
+    if len(co) == len(gb) == 0:
+        return cyc
+
+    input_data = cyc.by_key()
+    output_data = [input_data[i] for i in io + co]
+
+    for parent, children in gb.items():
+        if parent in co:
+            raise ValueError("complex device came in whole and in parts")
```
Comment (danielballan) [inline bluesky/utils.py]: This seems to invite a confused bug report. Can we expand on this a bit?

**Comment by tacaswell** [comment]
This will rely on https://github.com/NSLS-II/ophyd/pull/461 to work

File: test-requirements.txt [hunk 5]
```diff
@@ -10,7 +10,7 @@ matplotlib
 mongoquery
 networkx
 nose
-ophyd
+git+git://github.com/tacaswell/ophyd.git@enh_pseudopos_extensions#egg=ophyd
```
Comment (danielballan) [inline test-requirements.txt]: Do you plan to blow this commit away before merging? I would rather take your word that this combination of branches pass or, at most, aim this as ophyd master.

Comment (tacaswell) [reply test-requirements.txt]: Yes, the ophyd one needs to go in first.  I want it to run on travis so I can look at the coverage.

**Comment by tacaswell** [comment]
I think this one and the companion ophyd PR are ready for review.

The ophyd PR should go in first so we can adjust the testing on this to not look at a branch on my fork.

**Comment by tacaswell** [comment]
@ambarb This should fix your h vs k scans.

**Comment by tacaswell** [comment]
I think there is a bug in codecov where it is getting confused by `bluesk/utils.py` vs `blueskys/tests/utils.py`....

**Comment by ambarb** [comment]
@tacaswell do you mean that I should now be able to 
`RE(relative_scan(dets, tardis.h,-.1,0.1,20, tardis.k,-.1,0.1,20, tardis.l,-.1,0.1,20)`

or are you talking about some kind of plot?

**Comment by danielballan** [comment]
That’s exactly what we are after here. No more need for hacked solution here.

Get Outlook for iOS<https://aka.ms/o0ukef>



On Thu, Oct 26, 2017 at 5:31 PM -0400, "Andi Barbour" <notifications@github.com<mailto:notifications@github.com>> wrote:


@tacaswell<https://github.com/tacaswell> do you mean that I should now be able to
RE(relative_scan(dets, tardis.h,-.1,0.1,20, tardis.k,-.1,0.1,20, tardis.l,-.1,0.1,20)

or are you talking about some kind of plot?

**Comment by ambarb** [comment]
I wish markdown had `:Minions_exclaiming_YAY:`. Only that can express my enthusiasm.  

Do we want to try next week if you have not yet already?  I think this will cover about 85% of what people want to do in reciprocal space.

**Comment by tacaswell** [comment]
hmm, did not test the _relative_ case....

**Comment by ambarb** [comment]
By default, spec cannot do relative either.

I wrote a generic scan.  So maybe I should be specific and not be lazy with all the typing.

We get `inner_product_scan`and `outer_product_scan`?  We found this problem back with `a3scan` so that is why i left off the stuff in front of the scan.

**Comment by tacaswell** [comment]
I'll see if I can make it work, I do not want to make the minions sad.

**Comment by danielballan** [comment]
Wait.... who are the minions? This whole time I thought we were the minions.

Get Outlook for iOS<https://aka.ms/o0ukef>



On Thu, Oct 26, 2017 at 8:18 PM -0400, "Thomas A Caswell" <notifications@github.com<mailto:notifications@github.com>> wrote:


I'll see if I can make it work, I do not want to make the minions sad.

**Comment by ambarb** [comment]
`:more_happy_Minions:`
Would that mean it works with `scan_nd()`?  Sorry http://nsls-ii.github.io/bluesky/_sources/generated/bluesky.plans.relative_outer_product_scan.rst.txt  is broken. 

If so, then we can abandon calculating the motor positions for this case. scan_nd() is used for:
* fixed HKL energy scans 
* scanning a reciprocal space circle (more of a hard x-ray thing for aligning crystals)
* maybe someone would want to use a spiral scan instead of a mesh (since bp.spirals all rely on scan_nd) but maybe this last one is a bit crazy.

Then that would cover everything, unless there is a demand for `facility_circle_scan()` (spec has) or `facility_fixed_q_escan()` (spec doesn't have (to my knowledge))  Maybe these are wishlist items though?

**Comment by ambarb** [comment]
https://www.youtube.com/watch?v=cahV83nd7_I

**Comment by tacaswell** [comment]
Yeah, it is silently broken.... :disappointed:

**Comment by tacaswell** [comment]
@ambarb Still need to test it a bit more throughly, but _now_ it should work!

**Comment by ambarb** [comment]
@tacaswell awesome news.   I see that it was merged today.  Could we get an update to try it out?  There is an opening on Wednesday

**Comment by danielballan** [comment]
Yes. We're tagging ophyd v0.8.0 and bluesky v0.11.0 tonight.

**Pull request merged**